node/src/components/contract_runtime.rs

```diff
@@ -726,411 +541,79 @@ impl ContractRuntime {
-                                "deploy for block in era={} and height={} is expected to exist \
-                                in the storage",
-                                era_id, height
-                            )
-                            })
-                        })
-                        .collect(),
-                }))
-            })
-    }
+        deploys: Vec<Deploy>,
+    ) -> Effects<ContractRuntimeRequest>
+    where
+        REv: From<ContractRuntimeRequest>
+            + From<ContractRuntimeAnnouncement>
+            + From<ControlAnnouncement>
+            + Send,
+    {
+        let BlockAndExecutionEffects {
+            block,
+            execution_results,
+            maybe_step_execution_effect,
+        } = match operations::execute_finalized_block(
+            self.engine_state.as_ref(),
+            self.metrics.as_ref(),
+            protocol_version,
+            execution_pre_state,
+            finalized_block,
+            deploys,
+        ) {
+            Ok(block_and_execution_effects) => block_and_execution_effects,
+            Err(error) => return fatal!(effect_builder, "{}", error).ignore(),
+        };
 
-    /// Creates and announces the linear chain block.
-    fn finalize_block_execution<REv: ReactorEventT>(
-        &mut self,
-        effect_builder: EffectBuilder<REv>,
-        state: Box<RequestState>,
-        next_era_validator_weights: Option<BTreeMap<PublicKey, U512>>,
-    ) -> Effects<Event> {
-        // The state hash of the last execute-commit cycle is used as the block's post state
-        // hash.
-        let next_height = state.finalized_block.height() + 1;
-        // Update the metric.
-        self.metrics
-            .chain_height
-            .set(state.finalized_block.height() as i64);
-        let block = self.create_block(
-            state.finalized_block,
-            state.state_root_hash,
-            next_era_validator_weights,
-        );
+        self.execution_pre_state = ExecutionPreState::from(block.header());
+        let block_height = block.height();
 
+        let era_id = block.header().era_id();
         let mut effects = effect_builder
-            .announce_linear_chain_block(block, state.execution_results)
+            .announce_linear_chain_block(block, execution_results)
             .ignore();
+        if let Some(step_execution_effect) = maybe_step_execution_effect {
+            effects.extend(
+                effect_builder
+                    .announce_step_success(era_id, step_execution_effect)
+                    .ignore(),
+            );
+        }
+
         // If the child is already finalized, start execution.
-        if let Some((finalized_block, deploys)) = self.exec_queue.remove(&next_height) {
-            effects.extend(self.handle_get_deploys_result(
+        if let Some((finalized_block, deploys)) = self.exec_queue.remove(&(block_height + 1)) {
+            effects.extend(self.execute_finalized_block(
```

Comment:
@xcthulhu you mentioned this in our review of these changes. Does this mean a breaking change? Could you outline a bit about that?

Reply:
Fixed in https://github.com/casper-network/casper-node/pull/1669/commits/302374993792ed12d36a47094ae61e4cef65eae1

If we recur here, we'll never emit any effects until the queue is drained, which might mean downstream components will wait longer before they can do their thing.